usb: a few tweaks for battery powered applications

jfischer-no opened the pull request:
If the device controller can detect USB state changes, the application can enable/disable USB support on appropriate events. For the application in the tree, we need a mechanism to obtain this controller capability to keep the examples generic.
(Perhaps we should replace k_mutex with atomic in the stack exclusive access control to make the usbd_enable() call from the system workqueue context more robust.)

jfischer-no committed:
usb: device_next: tweak USB device notification support

Since no threads are waiting for the FIFO and we do not need any k_fifo
features, we can just use slist.

Some notifications (caused by interrupts), especially "VBUS ready", can
be triggered very quickly right after UDC initialization if the USB
controller is already connected. Depending on the global thread
configuration, messsage publishing work can be executed before the
context is marked as initialized. Use delayed work to reschedule
publishing when the context is not initialized.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: add USB device context to message delivery callback

Add USB device context argument to message delivery callback.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: remove initialized state checks in event processing

For the simple events, do not check whether the device driver and stack
are marked as initialized. USB device notification will reschedule
delivery if the stack is not yet marked initialized.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: add controller can detect VBUS state change capability

There are controllers with capability to detect VBUS state change. This
can be used in a generic application to explicitly handle the VBUS
detected event after usbd_init() and for example allow the PMIC to
detect to what type of port device controller is connected.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc_nrf: rework controller initialization

udc_init() only initializes and enables the USB regulator. This is
enough to detect VBUS changes, the USB controller is enabled in
udc_enable(). Deinitialization happens correspondingly in reverse order.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: remove unused API function

Remove unused and not required usbd_ep_ctrl_buf_alloc().

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: cdc_acm: enable USB support only when device is connected

If the controller can detect VBUS state changes, enable USB support only
when the device is connected and disable it when the device is
disconnected.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: cdc_acm: set USBD_CDC_ACM_LOG_LEVEL to error

Set USBD_CDC_ACM_LOG_LEVEL to error.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on subsys/usb/device_next/usbd_device.h:
```diff
@@ -79,9 +79,9 @@ static inline void usbd_status_suspended(struct usbd_contex *const uds_ctx,
+static inline int usbd_device_lock(struct usbd_contex *const uds_ctx)
 {
-	k_mutex_lock(&uds_ctx->mutex, K_FOREVER);
+	return k_mutex_lock(&uds_ctx->mutex, K_NO_WAIT);
```
Is this necessary? This puts quite big responsibility on checking for `-EBUSY` on the callers.

Thalley commented on include/zephyr/usb/usbd.h:
```diff
@@ -532,13 +547,15 @@ int usbd_init(struct usbd_contex *uds_ctx);
  *
- * This function enables the USB device support.
+ * This function enables the USB device support. If the wait argument is true,
+ * the stack waits for exclusive access, otherwise it does not wait and may
+ * return EBUSY if access is locked.
  *
  * @param[in] uds_ctx Pointer to USB device support context
  *
  * @return 0 on success, other values on fail.
  */
-int usbd_enable(struct usbd_contex *uds_ctx);
+int usbd_enable(struct usbd_contex *uds_ctx, const bool wait);
```
`wait` isn't documented properly as a parameters in doxygen. 

Why isn't the documentation build failing on this?

Thalley replied:
The `wait` parameter seems to always be set to true in samples and tests.

Is there an actual use case for it being `false`?

Thalley commented on include/zephyr/usb/usbd.h:
```diff
@@ -155,6 +155,21 @@ struct usbd_status {
+ * context of the system workqueue. Notification messages are stored in a queue
+ * and delivered to the callback in sequence.
+ *
+ * @param[in] msg Pointer to USB device message
+ */
+typedef void (*usbd_msg_cb_t)(struct usbd_contex *const ctx,
+			      const struct usbd_msg *const msg);
```
`ctx` is missing a `@param`

Thalley commented on subsys/usb/device_next/usbd_device.c:
```diff
@@ -253,3 +256,10 @@ int usbd_shutdown(struct usbd_contex *const uds_ctx)
+
+bool usbd_can_detect_vbus(struct usbd_contex *const uds_ctx)
+{
+	struct udc_device_caps caps = udc_caps(uds_ctx->dev);
```
```suggestion
	const struct udc_device_caps caps = udc_caps(uds_ctx->dev);
```

Thalley commented on tests/subsys/usb/device_next/src/main.c:
```diff
@@ -129,7 +129,7 @@ static void *usb_test_enable(void)
 	zassert_equal(err, 0, "Failed to initialize device support");
 
-	err = usbd_enable(&test_usbd);
+	err = usbd_enable(&test_usbd, true);
```
Should we have a test where `wait` is `false` as well? Otherwise it seems completely untested

jfischer-no replied:
There was a use in the sample, but I have dropped that commit.

jfischer-no committed:
usb: device_next: lock scheduler in usbd_init()

Lock the scheduler to ensure that the context is not preempted before it
is fully initialized.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented on subsys/usb/device_next/usbd_msg.c:
```diff
@@ -41,24 +43,47 @@ static inline void usbd_msg_pub(struct usbd_contex *const ctx,
+	k_spinlock_key_t key;
+	sys_snode_t *node;
+
+	key = k_spin_lock(&ml_lock);
+	node = sys_slist_peek_head(&msg_list);
+	k_spin_unlock(&ml_lock, key);
+
+	__ASSERT(node != NULL, "slist appears to be empty");
+	m_pkt = SYS_SLIST_CONTAINER(node, m_pkt, node);
```
I am quite sure I can, I do not need to protect it after peeking, I do not modify it and the only place where it can be removed is this block.

tmon-nordic replied:
The locking principle is violated regardless if the code works by coincidence or not.

tmon-nordic replied:
Note that all that I expect here is to just move `k_spin_unlock(&ml_lock, key);` to the point where `node` is no longer accessed.

Thalley commented on samples/subsys/usb/cdc_acm/src/main.c:
```diff
@@ -52,10 +52,24 @@ static inline void print_baudrate(const struct device *dev)
+			if (usbd_enable(ctx)) {
+				LOG_ERR("Failed to enable device support");
+			}
+		}
+
+		if (msg->type == USBD_MSG_VBUS_REMOVED) {
```
```suggestion
		} else if (msg->type == USBD_MSG_VBUS_REMOVED) {
```

Should there be an additional check/log statement if the type is neither of those types?

jfischer-no replied:
> Should there be an additional check/log statement if the type is neither of those types?

No, in this block we do not care about others.

tmon-nordic commented on subsys/usb/device_next/Kconfig:
```diff
@@ -47,6 +47,14 @@ config USBD_MSG_SLAB_COUNT
+	default 1
+	help
+	  Message work may need to be delayed because the device stack is not
+	  yet ready to publish the message. The delay unit is microseconds.
```
The config is used with `K_MSEC()` that defines the time in milliseconds not microseconds.

carlescufi commented:
@Thalley can you refresh your +1?
